Issue #445. Testing the behavior of the strncasecmp_custom function.

lecapitainenemo opened the pull request:
Related to issue #445 fix custom strncasecmp implementation.

It was described that the strncasecmp_custom function was going to fail and emit that strings like "not" and "NOTICE" were equal. However, the current implementation passed tests of this style (I added them).

This was a not a notable change so I did not add anything to ChangeLog.

Thanks.

DanD666 committed:
Added tests to test behavior of the strncasecmp_custom function. Related to issue #445 fix custon strncasecmp implementation.

lecapitainenemo commented:
I was wondering how to proceed, if I committed a mistake or if there is any other recommended task.
Thank you

goatshriek commented:
The bug can be subtle, but is present. By passing in a buffer size far greater than the string you provided, you've forced the comparison to go farther in and find the NULL character difference. However the from_buffer function needs to work when it doesn't have a NULL terminator. After all the purpose of this function is to NOT need a NULL terminator to begin with!

You can trigger the failure by reducing the buffer size to match the string length, add this test, or [my recommendation] do both.

```cpp
TEST( GetSeverityEnum, InvalidButPrefix ) {
  enum stumpless_severity result = stumpless_get_severity_enum( "war" );
  EXPECT_EQ( result, -1 );
  EXPECT_ERROR_ID_EQ( STUMPLESS_INVALID_SEVERITY );
}
```

DanD666 committed:
On issue #445. Two failing tests where added to severity.cpp. After that, the strncasecmp_custom function was modified to obtain the desired behavior.

lecapitainenemo commented:
Hey,
I added the failing tests. I removed the "EXPECT_ERROR_ID_EQ( STUMPLESS_INVALID_SEVERITY );" part from them, though. With the modification to strncasecmp_custom function, all the tests (new and old) pass. Please let me know if there is anything wrong. I appreciate your time and patience.

goatshriek commented:
Ah, I didn't realize that error codes were not being raised. I've opened #455 to address that.

Everything looks good! One final ask: can you also add a new test for the opposite direction: that strings that start with a valid severity but are longer are not accepted. For example "warnings are neat" should not be parsed to a valid severity. This will make sure that the comparison works properly in both directions.

DanD666 committed:
Test were added for strncasecmp_custom. In this case, Overextended severity strings (Longer strings that contained valid severities in them) were tested.

lecapitainenemo commented:
The respective tests were added and all old and new passed.
Thank you

goatshriek commented:
Thanks very much for putting this contribution together and sticking with it through the requested changes!
